### PR TITLE
#531: Use an explicit condition rather than case-when-bool

### PR DESCRIPTION
**Pull request by rbetzler**
Here's my attempt at #531. This is my first go at a SQLFluff PR, so hopefully I'm not too far off the mark 😅 

Thanks in advance for taking the time to review. Looking forward to get more involved in the project.

### CODE REVIEW

File: src/sqlfluff/core/rules/std/L043.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        from fancy_table
+
+    """
+
+    def _eval(self, segment, **kwargs):
+        """Find rule violations and provide fixes.
+
+        0. Look for a case expression
+        1. Find the first expression and "then"
+        2. Determine if "then" is followed by a boolean
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std/L043.py:35]: IIUC, the code checks for a Boolean after `THEN`, and if it finds one, it assumes the rest of the expression is:
```
else <<opposite value>>
```

What happens if this is not the case, e.g. if there's no else (result will be `NULL`, I think), or if it's `else <<complex expression>>>`?

In those cases, could we detect it doesn't fit the pattern and do nothing?

Comment (rbetzler) [reply src/sqlfluff/core/rules/std/L043.py:35]: Yeah that's a good callout. Will narrow this rule down to only handle: `case when... then bool else bool end`.

Along these lines, now that I think about it a little more, we might also need to wrap the fixed expression in a `coalesce`.

Will try to make these updates later today.

Comment (barrywhart) [reply src/sqlfluff/core/rules/std/L043.py:35]: Sounds great!

File: src/sqlfluff/core/rules/std/L043.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            case
+                when fab > 0 then true else false end as is_fab
+        from fancy_table
+
+    | **Best practice**
+
+    .. code-block:: sql
+
+        select
+            fab > 0 as is_fab
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std/L043.py:25]: Should this example use `coalesce()`?

File: src/sqlfluff/core/rules/std/L043.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
+"""Implementation of Rule L043."""
+
+from sqlfluff.core.rules.base import BaseCrawler, LintFix, LintResult
+from sqlfluff.core.rules.doc_decorators import document_fix_compatible
+
+
+@document_fix_compatible
+class Rule_L043(BaseCrawler):
+    """Use an explicit condition rather than case when statement.
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std/L043.py:9]: Can you think of a more precise way to explain the recommended practice (mentioning that it's not just "don't use case", but also about what the "case" statement is doing.

File: src/sqlfluff/core/rules/std/L043.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                else_bool_type = None
+                while segment.segments[idx].name != "END":
+                    if segment.segments[idx].raw_upper in ["TRUE", "FALSE"]:
+                        else_bool_type = segment.segments[idx].raw_upper
+                    idx += 1
+            # If then-bool-else-bool, return fixes
+            if (
+                then_bool_type is not None
+                and else_bool_type is not None
+                and then_bool_type != else_bool_type
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std/L043.py:77]: In order to test this line, can you add a "passing" test case(s) like this?
```
case
            when fab > 0 then true else true end
```

**Review by barrywhart — CHANGES_REQUESTED** [review]
The rule code looks 💯! Thanks for the updates. I have a few suggestions related to wording and adding one more test case. This is really close.

File: src/sqlfluff/core/rules/std/L043.py (hunk 3 above)
Comment (rbetzler) [reply src/sqlfluff/core/rules/std/L043.py]: Ah yeah that makes sense. I rewrote this a little. Hopefully it's clearer but not too declarative either.

**Review by barrywhart — APPROVED** [review]
Looks great!

**Comment by rbetzler** [comment]
@barrywhart, thank you for the reviews! Much appreciated 😄 

Excited to be involved in the project.

**Comment by barrywhart** [comment]
Thanks for the great work! Yes, it's a fun, cool, and *useful* project. 🎉

**Comment by rbetzler** [comment]
@barrywhart, mind if I ask: is there anything else you need on this one from me to get it merged? Not sure if you need to coordinate other PRs or need another reviewer before this one goes in, but thought I'd ask. Thanks again!

**Comment by dmateusp** [comment]
The coalesce bit is a very good catch, I actually had to test it myself in Snowflake because I didn't know about this behavior:

```sql
with

vals as (
    select * from values (1), (null), (-1) as t(n)
)

select
    n,
    n > 0,
    case when n > 0 then true else false end,
    coalesce(n > 0, false)
from vals
```

|N  |N > 0|CASE WHEN N > 0 THEN TRUE ELSE FALSE END|COALESCE(N > 0, FALSE)|
|---|-----|----------------------------------------|----------------------|
|1  |TRUE |TRUE                                    |TRUE                  |
|   |     |FALSE                                   |FALSE                 |
|-1 |FALSE|FALSE                                   |FALSE                 |

Leaving it here for reference

I'm going to go ahead and merge it, thanks @rbetzler 🎉

**Pull request merged**